compiler/AST/ParamForLoop.cpp

```diff
@@ -61,14 +62,42 @@ BlockStmt* ParamForLoop::buildParamForLoop(VarSymbol* indexVar,
     stride = new SymExpr(new_IntSymbol(1));
   }
 
-  if (call && call->isNamed("chpl_build_bounded_range"))
+  if (call && call->isNamed("#"))
   {
-    low    = call->get(1)->remove();
-    high   = call->get(1)->remove();
+    count = new CallExpr("chpl_compute_count_param_loop", call->get(2)->remove());
+    call        = toCallExpr(call->get(1));
+  }
+
+  if (count != NULL && call && call->isNamed("chpl_build_low_bounded_range"))
+  {
+    low  = call->get(1)->remove();
+    high = new CallExpr("chpl_high_bound_count_for_param_loop", low->copy(), count);
+  }
+  else if (count != NULL && call && call->isNamed("chpl_build_high_bounded_range"))
+  {
+    high = call->get(1)->remove();
+    low  = new CallExpr("chpl_low_bound_count_for_param_loop", high->copy(), count);
+  }
+  else if (call && call->isNamed("chpl_build_bounded_range"))
+  {
+    if(count != NULL)
+    {
+      Expr* temp_low = call->get(1)->remove();
+      Expr* temp_high = call->get(1)->remove();
+
+      // It is necessary that low is calculated first, because it also applies check for bound size.
+      low = new CallExpr("chpl_bounded_count_for_param_loop_low", temp_low->copy(), temp_high->copy(), count->copy());
+      high = new CallExpr("chpl_bounded_count_for_param_loop_high", temp_low->copy(), temp_high->copy(), count);
+    }
+    else
+    {
+      low    = call->get(1)->remove();
+      high   = call->get(1)->remove();
+    }
   }
   else
   {
-    USR_FATAL(range, "iterators for param-for-loops must be bounded literal ranges");
+    USR_FATAL(range, "param for loops currently only support range expressions with well defined param integral bounds");
```

Comment:
Minor, but I'd hyphenate "for-loops" here ("param for-loops") in order to avoid having the user accidentally read "for" as if it were the English word rather than the Chapel keyword.  I'd also hyphenate "well-defined".

Reply:
Yes, sir. I will do this for all the errors.